### ISSUE TITLE
Separating AMCP responses on TCP socket not possible?

### ISSUE THREAD

**Issue by tadams42** [CasparCG/Server#616]
Hi there,

I've been playing with AMCP client in Python... I'm doing something like this:

- open TCP client socket
- send some AMCP command
- call `socket.recv` to get response

How do I know that I got whole response from the stream of bytes received? 

- There is no length header to tell how many octets to expect which would allow me to call `socket.recv` few times and know when I'd received whole response message.   
- there is no clear message separator that would allow me to know when octets of one response message stop and new ones start (ie. using `\r\n` is not suitable for this because messages can contain that sequence as part of message body)

Am I missing something?

Basically, I propose adding to protocol either of following:

- fixed 4 octets header that indicates length of message following it
- unique message separator sequence. Maybe `\r\n\r\n` could be used because I think no responses contain this sequence.

**Comment by jesperstarkar** [comment]
It is indeed possible. The «Return code» section of the wiki describes the system: http://casparcg.com/wiki/CasparCG_2.1_AMCP_Protocol#LOADBG

You can have a look at SuperFlyTV/casparcg-connection how I have implemented it in Javascript. It might not be optimized, but it is solid and handles all scenarios + error handling.  The algoritm is in src/lib/CasparCGSocket.ts -> _parseResponseGroups(). This method receives strings separated by \r\n. 

As this is not an issue with CasparCG Server, but a question, I suggest this issue to be closed. Good luck.

**Comment by tadams42** [comment]
@jesperstarkar, thnx, I'll check your implementation.
What got me worried was that response codes come at the start of the message, meaning I don't know I'd reached the end of it until next one starts. Which basically answers my question for most cases. I guess I can introduce timeout on reading that way handling the last response in buffer. 

Anyway, closing this, thnx again.

**Comment by jesperstarkar** [comment]
\r\n marks the end. You can indeed separste exact where a response ends. That’s how all the other clients and libraries does.
